<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>Shane Hogle</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="Microbes + Biogeochemistry">
    <meta name="author" content="Shane Hogle">
    
    <link rel="canonical" href="//OTU_pipeline/">

    <!-- Custom CSS -->
    <link rel="stylesheet" href="/css/pixyll.css" type="text/css">

    <!-- Fonts -->
    <link href='//fonts.googleapis.com/css?family=Merriweather:900,900italic,300,300italic' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Lato:900,300' rel='stylesheet' type='text/css'>
    

    <!-- Open Graph -->
    <!-- From: https://github.com/mmistakes/hpstr-jekyll-theme/blob/master/_includes/head.html -->
    <meta property="og:locale" content="en_US">
    <meta property="og:type" content="article">
    <meta property="og:title" content="Shane Hogle">
    <meta property="og:description" content="Microbes + Biogeochemistry">
    <meta property="og:url" content="//OTU_pipeline/">
    <meta property="og:site_name" content="Shane Hogle">
</head>

<body class="">
  <div class="site-wrap">
    <header class="site-header px2 px-responsive">
  <div class="mt2 wrap">
    <div class="measure">
      <a href="/" class="site-title">Shane Hogle</a>
      <nav class="site-nav right">
        <a href="/about/">About</a>
<a href="/publications/">Publications</a>
<a href="/research/">Research</a>
<a href="/contact/">Contact</a>

      </nav>
      <div class="clearfix"></div>
      
    </div>
  </div>
</header>


    <div class="post p2 p-responsive wrap" role="main">
      <div class="measure">
        <div class="post">
  <header class="post-header">
    <h1 class="h2"></h1>
  </header>
  <article class="post-content">
  <h3 id="introduction">Introduction</h3>

<p>The OTU processing pipeline implemented here is basically the same as the ones recommended by Robert Edgar <a href="http://drive5.com/usearch/manual/uparse_pipeline.html">here</a> and <a href="http://drive5.com/usearch/manual/upp_ill_pe.html">here.</a> The only difference is that it doesn&rsquo;t use USEARCH to assign taxonomy. To run this pipeline you will need the <a href="http://www.drive5.com/usearch/download.html">USEARCH</a> software which includes the UPARSE algorithm.</p>

<p><a href="http://drive5.com/python/">Here</a> is a collection of Python scripts useful for working with USEARCH</p>

<h3 id="read-quality-control">Read quality control</h3>

<p>Merge paired end reads, allowing no differences in the overlap region.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -fastq_mergepairs DNA8_S22_L001_R1_001.fastq <span class="se">\</span>
-reverse DNA8_S22_L001_R2_001.fastq <span class="se">\</span>
-fastq_maxdiffs 0
-fastqout DNA7_merged.fastq
</code></pre></div>
<p>Get stats on the quality of the reads</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -fastq_stats DNA07_merged.fastq -log stats.log
</code></pre></div>
<p>Quality filter. Truncate reads at 250 bp and at a quality of 15.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -fastq_filter DNA07_merged.fastq <span class="se">\</span>
-fastq_trunclen <span class="m">250</span> <span class="se">\</span>
-fastq_truncqual <span class="m">15</span> <span class="se">\</span>
-fastaout DNA7_merged.filtered.fasta 
</code></pre></div>
<p>sed command to add barcodelabel tag for downstream demultiplexing. USEARCH cannot parse whitespace</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">sed <span class="s1">&#39;s/^&gt;\(.*\)/&gt;\1;barcodelabel=DNA8/&#39;</span> <span class="se">\</span>
DNA8_filtered.fa <span class="p">|</span> sed <span class="s1">&#39;s/\s//&#39;</span> &gt; DNA8.fa
</code></pre></div>
<p>Concatenate all the sample files together into one big file</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">cat *.fa &gt; allsamples.fa
</code></pre></div>
<h3 id="read-dereplication-step">Read dereplication step</h3>

<p>The input sequences to &rsquo;-cluster_otus&rsquo; must be a set of unique sequences sorted in order of decreasing abundance with size annotations in the labels. The &rsquo;-derep_fulllength&rsquo; command finds the unique sequences and add the size annotaitions.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -derep_fulllength allsamples.fa <span class="se">\ </span>
-relabel Uniq <span class="se">\</span>
-sizeout <span class="se">\ </span> 
-fastaout allsamples_derep.fa
</code></pre></div>
<h3 id="otu-clustering-step">OTU clustering step</h3>

<p>This command will sort the OTUs by size cluster size and discard singleton sequences</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -sortbysize allsamples_derep.fa <span class="se">\</span>
-minsize <span class="m">2</span> <span class="se">\</span>
-output allsamples_sorted.fa 
</code></pre></div>
<p>This command will cluster the sequences into OTUs</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -cluster_otus allsamples_sorted.fa <span class="se">\</span>
-otus allsamples_otus.fa
</code></pre></div>
<p>Now we can remove predicted chimeric sequences using a certified reference database. In this case we use the <a href="http://www.mothur.org/w/images/2/21/Greengenes.gold.alignment.zip">UCHIME gold standard reference.</a> In the <a href="http://drive5.com/usearch/manual/upp_ill_pe.html">UPARSE tutorial</a> this is ommitted. Perhaps newer versions of USEARCH do this automatically.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -uchime_ref allsamples_otus.fa <span class="se">\</span>
-db ./seq_datases/gold.fa <span class="se">\</span>
-strand plus <span class="se">\</span>
-nonchimeras allsamples_otus_chimslay.fa
</code></pre></div>
<p>Label OTU sequences OTU<em>1, OTU</em>2&hellip; The python scripts for this are the same referenced above.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">python ./python_scripts/fasta_number.py <span class="se">\</span>
allsamples_otus_chimslay.fa OTU_ &gt; allsamples_otus_final.fa
</code></pre></div>
<p>Map reads (including singletons) back to OTUs</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">usearch -usearch_global allsamples.fa <span class="se">\</span>
-db allsamples_otus_final.fa <span class="se">\</span>
-strand plus <span class="se">\</span>
-id 0.97 <span class="se">\</span>
-uc allsamples_map.uc
</code></pre></div>
<p>Now create an OTU table. The python scripts for this are the same referenced above.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">python ./python_scripts/uc2otutab.py <span class="se">\</span>
allsamples_map.uc &gt; allsamples_otu_table.txt
</code></pre></div>
  </article>
</div>

      </div>
    </div>
  </div>

  <footer class="footer">
  <div class="p2 wrap">
    <div class="measure mt1 center">
      <small>
        Â©Copyright 2014 Shane Hogle</a> (<a href="https://twitter.com/slhogle">@slhogle</a>).
      </small>
    </div>
  </div>
</footer>



</body>
</html>
