<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>Shane Hogle</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="Microbes + Biogeochemistry">
    <meta name="author" content="Shane Hogle">
    
    <link rel="canonical" href="//CCE2012_MAW_microbes/data_analysis/workflow/">

    <!-- Custom CSS -->
    <link rel="stylesheet" href="/css/pixyll.css" type="text/css">

    <!-- Fonts -->
    <link href='//fonts.googleapis.com/css?family=Merriweather:900,900italic,300,300italic' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Lato:900,300' rel='stylesheet' type='text/css'>
    

    <!-- Open Graph -->
    <!-- From: https://github.com/mmistakes/hpstr-jekyll-theme/blob/master/_includes/head.html -->
    <meta property="og:locale" content="en_US">
    <meta property="og:type" content="article">
    <meta property="og:title" content="Shane Hogle">
    <meta property="og:description" content="Microbes + Biogeochemistry">
    <meta property="og:url" content="//CCE2012_MAW_microbes/data_analysis/workflow/">
    <meta property="og:site_name" content="Shane Hogle">
</head>

<body class="">
  <div class="site-wrap">
    <header class="site-header px2 px-responsive">
  <div class="mt2 wrap">
    <div class="measure">
      <a href="/" class="site-title">Shane Hogle</a>
      <nav class="site-nav right">
        <a href="/about/">About</a>
<a href="/publications/">Publications</a>
<a href="/research/">Research</a>
<a href="/contact/">Contact</a>

      </nav>
      <div class="clearfix"></div>
      
    </div>
  </div>
</header>


    <div class="post p2 p-responsive wrap" role="main">
      <div class="measure">
        <div class="post">
  <header class="post-header">
    <h1 class="h2"></h1>
  </header>
  <article class="post-content">
  <h2 id="reference-publication">Reference publication</h2>

<p><strong>Hogle SL</strong>, Bundy RA, Barbeau KA. <em>In prep.</em> Linking phytoplankton and bacterioplankton assemblage dynamics to iron-binding ligand production in a microcosm experiment. </p>

<hr>

<h2 id="introduction-otu-data-analysis">Introduction: OTU data analysis</h2>

<p>The analyses we&rsquo;re doing here are going to be using the R package &ldquo;phyloseq.&rdquo; There are some great tutorials on how to use phyloseq <a href="http://joey711.github.io/phyloseq/tutorials-index">here</a>. In addition, the data that we&rsquo;re using here has been generated by an OTU pipeline based mainly on <a href="http://drive5.com/uparse/">UPARSE</a>. There is a lot of great information at the UPARSE website. The pipeline that I followed to obtain the data files </p>

<h3 id="load-required-packages">Load required packages</h3>
<div class="highlight"><pre><code class="language-R" data-lang="R"><span class="kn">library</span><span class="p">(</span>phyloseq<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>ggplot2<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>reshape<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>grid<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>vegan<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>DESeq2<span class="p">)</span>
</code></pre></div>
<h3 id="check-package-versions">Check package versions</h3>
<div class="highlight"><pre><code class="language-R" data-lang="R">packageVersion<span class="p">(</span><span class="s">&quot;phyloseq&quot;</span><span class="p">)</span>
packageVersion<span class="p">(</span><span class="s">&quot;ggplot2&quot;</span><span class="p">)</span>
packageVersion<span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">[1] ‘1.12.2’
[1] ‘1.0.1’
[1] ‘1.8.2’
</code></pre></div>
<h3 id="import-the-dataset-and-combine-individual-sample-files">Import the dataset and combine individual sample files</h3>

<p>Set working directory</p>
<div class="highlight"><pre><code class="language-R" data-lang="R"><span class="kp">setwd</span><span class="p">(</span><span class="s">&quot;YOUR/WORKING/DIRECTORY&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Import individual OTU tables from each sample</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">DNA07 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA07_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
DNA08 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA08_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
DNA09 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA09_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
DNA10 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA10_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
DNA11 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA11_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
DNA12 <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span><span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;DNA12_otutab.txt&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> select <span class="o">=</span> <span class="o">-</span>taxonomy<span class="p">)</span>
</code></pre></div>
<p>Create a list of the tables</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">samples.list <span class="o">=</span> <span class="kt">list</span><span class="p">(</span>DNA07<span class="p">,</span> DNA08<span class="p">,</span> DNA09<span class="p">,</span> DNA10<span class="p">,</span> DNA11<span class="p">,</span> DNA12<span class="p">)</span>
</code></pre></div>
<p>use reduce + merge to bind all the tables together</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTU.tab <span class="o">=</span> <span class="kp">Reduce</span><span class="p">(</span><span class="kr">function</span><span class="p">(</span><span class="kc">...</span><span class="p">)</span> <span class="kp">merge</span><span class="p">(</span><span class="kc">...</span><span class="p">,</span> all<span class="o">=</span><span class="bp">T</span><span class="p">),</span> samples.list<span class="p">)</span>
</code></pre></div>
<p>convert NAs to zeroes in the dataframe</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTU.tab<span class="p">[</span><span class="kp">is.na</span><span class="p">(</span>OTU.tab<span class="p">)]</span> <span class="o">&lt;-</span> <span class="m">0</span>
</code></pre></div>
<p>save a table with the combined OTUs for the samples</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">write.table<span class="p">(</span>OTU.tab<span class="p">,</span> file<span class="o">=</span><span class="s">&quot;allsamples_otu_table&quot;</span><span class="p">,</span> quote <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> row.names <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Set row names to column 1 and remove the original values from that column</p>
<div class="highlight"><pre><code class="language-R" data-lang="R"><span class="kp">rownames</span><span class="p">(</span>OTU.tab<span class="p">)</span> <span class="o">&lt;-</span> OTU.tab<span class="p">[,</span><span class="m">1</span><span class="p">]</span>
OTU.tab <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>OTU.tab<span class="p">,</span> select <span class="o">=</span> <span class="o">-</span>OTU_ID<span class="p">)</span>
</code></pre></div>
<p>Load the previously saved OTU matrix if needed</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTU.tab <span class="o">&lt;-</span> <span class="kp">as.matrix</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;allsamples_otu_table&quot;</span><span class="p">,</span> row.names<span class="o">=</span><span class="m">1</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
</code></pre></div>
<p>Load and create a taxonomy matrix</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">taxmat <span class="o">&lt;-</span> <span class="kp">as.matrix</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;rdp_tax_table&quot;</span><span class="p">,</span> row.names<span class="o">=</span><span class="m">1</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
</code></pre></div>
<p>Load and create a metadata data frame</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">metamat <span class="o">&lt;-</span> <span class="kp">as.data.frame</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;allsamples.metadata&quot;</span><span class="p">,</span> row.names<span class="o">=</span><span class="m">1</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
</code></pre></div>
<h3 id="generate-phyloseq-object">Generate phyloseq object</h3>

<p>Here we’ll call some of phyloseq’s functions to genearate an object called cce12 that contains metadata, OTU, and taxonomy info.</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTU <span class="o">=</span> otu_table<span class="p">(</span>OTU.tab<span class="p">,</span> taxa_are_rows <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
TAX <span class="o">=</span> tax_table<span class="p">(</span>taxmat<span class="p">)</span>
SAMPDAT <span class="o">=</span> sample_data<span class="p">(</span>metamat<span class="p">)</span>
cce12 <span class="o">=</span> phyloseq<span class="p">(</span>OTU<span class="p">,</span> TAX<span class="p">,</span> SAMPDAT<span class="p">)</span>
</code></pre></div>
<h3 id="lets-take-a-look-at-some-summary-statistics">Lets take a look at some summary statistics</h3>

<p>Generate OTU sums across all samples and sort</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTUsums <span class="o">=</span> <span class="kt">data.frame</span><span class="p">(</span>nreads <span class="o">=</span> <span class="kp">sort</span><span class="p">(</span>taxa_sums<span class="p">(</span>cce12<span class="p">),</span> <span class="kc">TRUE</span><span class="p">),</span> sorted <span class="o">=</span> <span class="m">1</span><span class="o">:</span>ntaxa<span class="p">(</span>cce12<span class="p">))</span>
</code></pre></div>
<p>Now calculate the number of reads in each sample and make a table</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">samplesums <span class="o">=</span> <span class="kt">data.frame</span><span class="p">(</span>read.count <span class="o">=</span> <span class="kp">sort</span><span class="p">(</span>sample_sums<span class="p">(</span>cce12<span class="p">),</span> <span class="kc">TRUE</span><span class="p">))</span>
</code></pre></div>
<p>Most of the samples are pretty evenly sequenced with the exception of DNA08. Now plot an OTU rank abundance curve using ggplot. We can see that there are many OTUs with a very small abundance</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">ggplot<span class="p">(</span>OTUsums<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> sorted<span class="p">,</span> y <span class="o">=</span> nreads<span class="p">))</span> <span class="o">+</span> geom_line<span class="p">(</span>stat <span class="o">=</span> <span class="s">&quot;identity&quot;</span><span class="p">,</span> size <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  scale_y_log10<span class="p">()</span> <span class="o">+</span> labs<span class="p">(</span>x <span class="o">=</span> <span class="s">&quot;OTU rank abundance&quot;</span><span class="p">,</span> y <span class="o">=</span> <span class="s">&quot;Total number of reads&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_41_0.svg" alt="svg"></p>

<h3 id="preprocessing-data">Preprocessing data</h3>

<p>Here we’ll do some preprocessing to remove spurious and nonabundant OTUs that we don’t want to affect downstream analysis</p>

<p>Here we use the subset taxa function to select only the Domain “Bacteria” while excluding “Cyanobacteria/chloroplast.” cce12_st is a new phyloseq object</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12_st <span class="o">&lt;-</span> subset_taxa<span class="p">(</span>cce12<span class="p">,</span> Phylum<span class="o">!=</span><span class="s">&quot;Cyanobacteria/Chloroplast&quot;</span><span class="p">)</span>
cce12_st <span class="o">&lt;-</span> subset_taxa<span class="p">(</span>cce12_st<span class="p">,</span> Domain<span class="o">==</span><span class="s">&quot;Bacteria&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Now we will transform the reads for each OTU in each sample to relative abundance - ie. reads for OTUx / total reads in sample x</p>

<p>This reuslts in 826 OTUs</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12st.rel.abun <span class="o">&lt;-</span> transform_sample_counts<span class="p">(</span>cce12_st<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">)</span> x <span class="o">/</span> <span class="kp">sum</span><span class="p">(</span>x<span class="p">))</span>
cce12st.rel.abun
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 826 taxa and 6 samples ]
sample_data() Sample Data:       [ 6 samples by 11 sample variables ]
tax_table()   Taxonomy Table:    [ 826 taxa by 6 taxonomic ranks ]
</code></pre></div>
<p>Now we will remove OTUs that have a mean percentage abundance of less than 0.005%. The resulting phyloseq object has 304 OTUs</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12st.rel.abun.filter <span class="o">&lt;-</span> filter_taxa<span class="p">(</span>cce12st.rel.abun<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">)</span> <span class="kp">mean</span><span class="p">(</span>x<span class="p">)</span> <span class="o">&gt;</span> <span class="m">5e-5</span><span class="p">,</span> <span class="kc">TRUE</span><span class="p">)</span>
cce12st.rel.abun.filter
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 304 taxa and 6 samples ]
sample_data() Sample Data:       [ 6 samples by 11 sample variables ]
tax_table()   Taxonomy Table:    [ 304 taxa by 6 taxonomic ranks ]
</code></pre></div>
<p>Collect OTUs from cce12.rel.abun.filter and use them to filter the raw OTUs from cce12</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12.raw.filt <span class="o">&lt;-</span> prune_taxa<span class="p">(</span>taxa_names<span class="p">(</span>cce12st.rel.abun.filter<span class="p">),</span> cce12<span class="p">)</span>
</code></pre></div>
<p>Remove taxa not seen more than 3 times in at least 20% of the samples. This protects against OTUs with small mean &amp; trivially large covariance</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12.raw.filt.2 <span class="o">&lt;-</span> filter_taxa<span class="p">(</span>cce12.raw.filt<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">)</span> <span class="kp">sum</span><span class="p">(</span>x <span class="o">&gt;</span> <span class="m">3</span><span class="p">)</span> <span class="o">&gt;</span> <span class="p">(</span><span class="m">0.2</span><span class="o">*</span><span class="kp">length</span><span class="p">(</span>x<span class="p">)),</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p>Generate OTU sums across all samples in the new filtered set and sort</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTUsums1 <span class="o">&lt;-</span> <span class="kt">data.frame</span><span class="p">(</span>nreads <span class="o">=</span> <span class="kp">sort</span><span class="p">(</span>taxa_sums<span class="p">(</span>cce12.raw.filt.2<span class="p">),</span> <span class="kc">TRUE</span><span class="p">),</span> sorted <span class="o">=</span> <span class="m">1</span><span class="o">:</span>ntaxa<span class="p">(</span>cce12.raw.filt.2<span class="p">))</span>
</code></pre></div>
<p>Plot new rank abundace curve based on reduced OTU set</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">ggplot<span class="p">(</span>OTUsums1<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> sorted<span class="p">,</span> y <span class="o">=</span> nreads<span class="p">))</span> <span class="o">+</span> geom_line<span class="p">(</span>stat <span class="o">=</span> <span class="s">&quot;identity&quot;</span><span class="p">,</span> size <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  scale_y_log10<span class="p">()</span> <span class="o">+</span> labs<span class="p">(</span>x <span class="o">=</span> <span class="s">&quot;OTU rank abundance&quot;</span><span class="p">,</span> y <span class="o">=</span> <span class="s">&quot;Total number of reads&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_56_0.svg" alt="svg"></p>

<p>Again make a summary table of read counts for each library</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">samplesums_postfilter <span class="o">&lt;-</span> <span class="kp">cbind</span><span class="p">(</span>samplesums<span class="p">,</span> 
                               <span class="kt">data.frame</span><span class="p">(</span>read.count.postfilter <span class="o">=</span> <span class="kp">sort</span><span class="p">(</span>sample_sums<span class="p">(</span>cce12.raw.filt<span class="p">),</span> <span class="kc">TRUE</span><span class="p">)),</span> 
                               <span class="kt">data.frame</span><span class="p">(</span>read.count.postfilter.2 <span class="o">=</span> <span class="kp">sort</span><span class="p">(</span>sample_sums<span class="p">(</span>cce12.raw.filt.2<span class="p">),</span> <span class="kc">TRUE</span><span class="p">)))</span>
samplesums_postfilter
</code></pre></div>
<p>Now we’ll standardize abundances to the median sequencing depth of each sample</p>

<p>First calculate the median for the sample sums in the filtered dataset</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">total.median <span class="o">&lt;-</span> median<span class="p">(</span>sample_sums<span class="p">(</span>cce12.raw.filt.2<span class="p">))</span>
</code></pre></div>
<p>Now create a function that standardized the relative abundance of each OTU to the median sequencing depth</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">stdize.fun <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">,</span> t<span class="o">=</span>total.median<span class="p">)</span> <span class="kp">round</span><span class="p">(</span>t <span class="o">*</span> <span class="p">(</span>x <span class="o">/</span> <span class="kp">sum</span><span class="p">(</span>x<span class="p">)))</span>
</code></pre></div>
<p>Create the standardized result</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">cce12norm <span class="o">=</span> transform_sample_counts<span class="p">(</span>cce12.raw.filt.2<span class="p">,</span> stdize.fun<span class="p">)</span>
</code></pre></div>
<h3 id="richness-alpha-diversity-of-data-partitioned-by-fe_treatment">Richness/alpha diversity of data partitioned by Fe_Treatment</h3>

<p>Here we’ll make some richness plots of our data separated by high iron and low iron treatments. We use raw data instead of the normalized data here because many richness estimates are based on singletons and doubletons. We need to leave them in order to get meaningful estimates.</p>

<p>Get numerical values for richness estimates and write output as a table</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">richness <span class="o">&lt;-</span> estimate_richness<span class="p">(</span>cce12_st<span class="p">)</span>
write.table<span class="p">(</span>richness<span class="p">,</span> file <span class="o">=</span> <span class="s">&quot;sample_richness_estimates.tsv&quot;</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">,</span> row.names <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre><code class="language-R" data-lang="R">p <span class="o">=</span> plot_richness<span class="p">(</span>cce12_st<span class="p">,</span> x <span class="o">=</span> <span class="s">&quot;Fe_Treatment&quot;</span><span class="p">,</span> color <span class="o">=</span> <span class="s">&quot;Sample_ID&quot;</span><span class="p">)</span>
p <span class="o">+</span> geom_point<span class="p">(</span>data <span class="o">=</span> p<span class="o">$</span>data<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> Fe_Treatment<span class="p">,</span> y <span class="o">=</span> value<span class="p">,</span> color <span class="o">=</span> Sample_ID<span class="p">),</span> size <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_69_0.svg" alt="svg"></p>

<p>Right now, we actually only care about richness plots of Chao1 and “observed” diversity so we’ll just select those</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">psub <span class="o">=</span> plot_richness<span class="p">(</span>cce12_st<span class="p">,</span> x <span class="o">=</span> <span class="s">&quot;Fe_Treatment&quot;</span><span class="p">,</span> color <span class="o">=</span> <span class="s">&quot;Sample_ID&quot;</span><span class="p">,</span> measures <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;Observed&quot;</span><span class="p">,</span> <span class="s">&quot;Chao1&quot;</span><span class="p">))</span>
psub <span class="o">+</span> geom_point<span class="p">(</span>data <span class="o">=</span> psub<span class="o">$</span>data<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> Fe_Treatment<span class="p">,</span> y <span class="o">=</span> value<span class="p">,</span> color <span class="o">=</span> Sample_ID<span class="p">),</span> size <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_71_0.svg" alt="svg"></p>

<h3 id="bar-plots">Bar plots</h3>

<p>Now let’s make some bar plots to get a fuller picture as to how the top OTUs are distributed throughout our Fe treatments. Here we’ll select the top 20 most abundant OTUs from the normalized dataset “cce12norm” and partition them by class.</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">topOTUs <span class="o">=</span> <span class="kp">names</span><span class="p">(</span><span class="kp">sort</span><span class="p">(</span>taxa_sums<span class="p">(</span>cce12norm<span class="p">),</span> <span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">])</span>
cce12norm20 <span class="o">=</span> prune_taxa<span class="p">(</span>topOTUs<span class="p">,</span> cce12norm<span class="p">)</span>
plot_bar<span class="p">(</span>cce12norm20<span class="p">,</span> <span class="s">&quot;Sample_ID&quot;</span><span class="p">,</span> fill <span class="o">=</span> <span class="s">&quot;Family&quot;</span><span class="p">,</span> facet_grid <span class="o">=</span> <span class="o">~</span>OTU<span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_74_0.svg" alt="svg"></p>

<p>To make bar plots of all OTUs combined into a particular taxonomic rank you need to use the tax_glom function.</p>

<p>Combine OTUs at the taxonomic rank of Class. Results in 25 OTUs</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">glommed <span class="o">&lt;-</span> tax_glom<span class="p">(</span>cce12norm<span class="p">,</span> taxrank<span class="o">=</span><span class="s">&quot;Class&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Create variable “OTUS_remove” that contains the names of the 19 least abundant OTUs in the glommed dataset</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTUS_remove <span class="o">&lt;-</span> <span class="kp">names</span><span class="p">(</span><span class="kp">sort</span><span class="p">(</span>taxa_sums<span class="p">(</span>glommed<span class="p">),</span> <span class="kc">FALSE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">])</span>
</code></pre></div>
<p>Remove the 19 least abundant Class &ldquo;OTUs&rdquo; from the glommed dataset</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">glommed.red <span class="o">&lt;-</span> merge_taxa<span class="p">(</span>glommed<span class="p">,</span> OTUS_remove<span class="p">,</span> <span class="m">1</span><span class="p">)</span>
</code></pre></div>
<p>This is kind of annoying&hellip;.</p>

<p>When you make the bar plot using the phyloseq wrapper &ldquo;plot_bar&rdquo; it puts the components not in the same order for each sample. I don’t actually understand how this wrapper is implemented in phyloseq</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">p <span class="o">&lt;-</span> plot_bar<span class="p">(</span>glommed.red<span class="p">,</span> <span class="s">&quot;Sample_ID&quot;</span><span class="p">,</span> fill <span class="o">=</span> <span class="s">&quot;Class&quot;</span><span class="p">)</span>
p <span class="o">+</span> geom_bar<span class="p">(</span>aes<span class="p">(</span>fill<span class="o">=</span><span class="s">&quot;Class&quot;</span><span class="p">,</span> position<span class="o">=</span><span class="s">&quot;stack&quot;</span><span class="p">))</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_82_0.svg" alt="svg"></p>

<p>Ok to make this not a jumbled bar plot mess I&rsquo;ll do this with vanilla ggplot2. First, set up data frames produced from the tax<em>table and otu</em>tables from the glommed.red dataset</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">OTUs.class <span class="o">&lt;-</span> <span class="kt">data.frame</span><span class="p">(</span>otu_table<span class="p">(</span>glommed.red<span class="p">))</span>
tax.class <span class="o">&lt;-</span> <span class="kt">data.frame</span><span class="p">(</span>tax_table<span class="p">(</span>glommed.red<span class="p">))</span>
</code></pre></div>
<p>Merge these data frames into one</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">merged.class.plot <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>OTUs.class<span class="p">,</span> tax.class<span class="p">,</span> by<span class="o">=</span><span class="s">&quot;row.names&quot;</span><span class="p">,</span> all<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
merged.class.plot
</code></pre></div>
<p>We&rsquo;ll now get rid of unecessary columns</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">merged.class.plot.red <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>merged.class.plot<span class="p">,</span> select<span class="o">=-</span><span class="kt">c</span><span class="p">(</span>Row.names<span class="p">,</span> Domain<span class="p">,</span> Phylum<span class="p">,</span> Order<span class="p">,</span> Family<span class="p">,</span> Genus<span class="p">))</span>
</code></pre></div>
<p>Now we&rsquo;ll put df into long format and select id.vars as &ldquo;Class&rdquo; using the reshape package</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">merged.class.plot.red.m <span class="o">&lt;-</span> melt<span class="p">(</span>merged.class.plot.red<span class="p">,</span> id.vars<span class="o">=</span><span class="s">&quot;Class&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Finally now we&rsquo;ll make the bar plot with the correct ordering of the different classes, which looks a lot nicer.</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">ggplot<span class="p">(</span>merged.class.plot.red.m<span class="p">,</span> aes<span class="p">(</span>variable<span class="p">,</span> value<span class="p">,</span> fill<span class="o">=</span>Class<span class="p">))</span> <span class="o">+</span> geom_bar<span class="p">(</span>stat<span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_92_0.svg" alt="svg"></p>

<h3 id="detecting-otus-with-differential-abundance-between-fe-treatments">Detecting OTUs with differential abundance between Fe Treatments</h3>

<p>So there seems to be some controvery in the microbiome world about whether or not to Rarefy data. Rarefying is basically a library size normalization technique that randomly subsamples reads from each library/sample in order to make all libraries contain the same number fo reads as the smallest library. This is the default normalization proceedure in QIIME as far as I can tell. A <a href="https://peerj.com/preprints/1157/">very recent paper</a> states that Rarefying is a good option when your library sizes are vastly different, but if they are close to each other in read number (like ours) and you have a small sample size (also like ours) then procedures utilizing generalized linear models (like DESeq2) to be more sensitive and powerful.</p>

<p>Some studies have gone as far as saying that Rarefying is <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">&ldquo;inexcusable&rdquo;</a>. It seems like the best option for us is to model microbiome counts with a Negative Binomial model as implemented in the popular package <a href="http://www.genomebiology.com/2014/15/12/550">DESeq2</a> for detecting differential expression in RNAseq data. Here we use DESeq2 to test whether any particular OTUs are significantly less or more abundant in high iron versus low iron treatments.</p>

<p>Import data with phyloseq, convert to DESeq2’s DESeqDataSet class</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">FeTreat.dds <span class="o">=</span> phyloseq_to_deseq2<span class="p">(</span>cce12.raw.filt.2<span class="p">,</span> <span class="o">~</span> Fe_Treatment<span class="p">)</span>
</code></pre></div>
<p>Calculate geometric means prior to estimate size factors</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">gm_mean <span class="o">=</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">,</span> na.rm<span class="o">=</span><span class="kc">TRUE</span><span class="p">){</span>
  <span class="kp">exp</span><span class="p">(</span><span class="kp">sum</span><span class="p">(</span><span class="kp">log</span><span class="p">(</span>x<span class="p">[</span>x <span class="o">&gt;</span> <span class="m">0</span><span class="p">]),</span> na.rm<span class="o">=</span>na.rm<span class="p">)</span> <span class="o">/</span> <span class="kp">length</span><span class="p">(</span>x<span class="p">))</span>
<span class="p">}</span>
geoMeans <span class="o">=</span> <span class="kp">apply</span><span class="p">(</span>counts<span class="p">(</span>FeTreat.dds<span class="p">),</span> <span class="m">1</span><span class="p">,</span> gm_mean<span class="p">)</span>
FeTreat.dds <span class="o">=</span> estimateSizeFactors<span class="p">(</span>FeTreat.dds<span class="p">,</span> geoMeans <span class="o">=</span> geoMeans<span class="p">)</span>
</code></pre></div>
<p>Running DESeq2 on non transformed data</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">FeTreat.dds <span class="o">=</span> DESeq<span class="p">(</span>FeTreat.dds<span class="p">,</span> fitType<span class="o">=</span><span class="s">&quot;local&quot;</span><span class="p">)</span>
res <span class="o">=</span> results<span class="p">(</span>FeTreat.dds<span class="p">)</span>
res <span class="o">=</span> res<span class="p">[</span><span class="kp">order</span><span class="p">(</span>res<span class="o">$</span>padj<span class="p">,</span> na.last<span class="o">=</span><span class="kc">NA</span><span class="p">),</span> <span class="p">]</span>
alpha <span class="o">=</span> <span class="m">0.05</span>
sigtab <span class="o">=</span> res<span class="p">[(</span>res<span class="o">$</span>padj <span class="o">&lt;</span> alpha<span class="p">),</span> <span class="p">]</span>
sigtab <span class="o">=</span> <span class="kp">cbind</span><span class="p">(</span>as<span class="p">(</span>sigtab<span class="p">,</span> <span class="s">&quot;data.frame&quot;</span><span class="p">),</span> as<span class="p">(</span>tax_table<span class="p">(</span>cce12<span class="p">)[</span><span class="kp">rownames</span><span class="p">(</span>sigtab<span class="p">),</span> <span class="p">],</span> <span class="s">&quot;matrix&quot;</span><span class="p">))</span>
write.table<span class="p">(</span>sigtab<span class="p">,</span> file <span class="o">=</span> <span class="s">&quot;diff_abun_OTU.tsv&quot;</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">,</span> row.names <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
sigtab_red <span class="o">=</span> <span class="kp">subset</span><span class="p">(</span>sigtab<span class="p">,</span> select <span class="o">=</span> <span class="o">-</span><span class="kt">c</span><span class="p">(</span>Domain<span class="p">,</span> Phylum<span class="p">,</span> Order<span class="p">,</span> Family<span class="p">))</span>
</code></pre></div>
<p>cool - so it looks like there are 9 OTUs that are significantly more abundant in high Fe treatments (those with a negative log2 Fold-change) and 13 that are less abundant.</p>

<h3 id="ordination-of-data-using-phyloseq-vegan-ggplot-and-some-other-stuff…">Ordination of data using phyloseq, vegan, ggplot and some other stuff…</h3>

<p>Finally we’ll make an ordination based on our community data and fit continuous environmental variable to it as factors</p>

<p>First define a function veganotu for extracting an OTU table from a phyloseq object and coercing it into a form vegan likes</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">veganotu <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>physeq<span class="p">)</span> <span class="p">{</span>
    OTU <span class="o">&lt;-</span> otu_table<span class="p">(</span>physeq<span class="p">)</span>
    <span class="kr">if</span> <span class="p">(</span>taxa_are_rows<span class="p">(</span>OTU<span class="p">))</span> <span class="p">{</span>
        OTU <span class="o">&lt;-</span> <span class="kp">t</span><span class="p">(</span>OTU<span class="p">)</span>
    <span class="p">}</span>
    <span class="kr">return</span><span class="p">(</span>as<span class="p">(</span>OTU<span class="p">,</span> <span class="s">&quot;matrix&quot;</span><span class="p">))</span>
<span class="p">}</span>
</code></pre></div>
<p>Convert physeq metadata into vanilla R data frames</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">keepvariables <span class="o">=</span> <span class="kp">which</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>sample_data<span class="p">(</span>cce12.raw.filt.2<span class="p">),</span> <span class="kp">is.numeric</span><span class="p">))</span>
physeqsd <span class="o">=</span> <span class="kt">data.frame</span><span class="p">(</span>sample_data<span class="p">(</span>cce12.raw.filt.2<span class="p">))[</span>keepvariables<span class="p">]</span>
physeqsd <span class="o">=</span> <span class="kp">subset</span><span class="p">(</span>physeqsd<span class="p">,</span> select <span class="o">=</span> <span class="o">-</span>Fe_level<span class="p">)</span>
treatment <span class="o">=</span> <span class="kt">data.frame</span><span class="p">(</span>sample_data<span class="p">(</span>cce12.raw.filt.2<span class="p">))[</span><span class="s">&quot;Fe_Treatment&quot;</span><span class="p">]</span>
</code></pre></div>
<p>Using the normalized OTU matrix in the metaMDS function</p>
<div class="highlight"><pre><code class="language-R" data-lang="R"><span class="kp">set.seed</span><span class="p">(</span><span class="m">4019</span><span class="p">)</span>
vare.mds <span class="o">&lt;-</span> metaMDS<span class="p">(</span>veganotu<span class="p">(</span>cce12norm<span class="p">),</span> trace <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> plot <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> autotransform <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> noshare <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> wascores <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> trymax <span class="o">=</span> <span class="m">1000</span><span class="p">)</span>
</code></pre></div>
<p>Use the score function to load sample sites into vectors. We also use plot_ordination function with the justDF option =TRUE in order to load the specific coordinates of just the OTUs.</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">scrs.site <span class="o">&lt;-</span> <span class="kp">as.data.frame</span><span class="p">(</span>scores<span class="p">(</span>vare.mds<span class="p">,</span> display <span class="o">=</span> <span class="s">&quot;sites&quot;</span><span class="p">))</span>
scrs.site <span class="o">&lt;-</span> <span class="kp">cbind</span><span class="p">(</span>scrs.site<span class="p">,</span> treatment <span class="o">=</span> treatment<span class="p">)</span>
scrs.spp <span class="o">&lt;-</span> plot_ordination<span class="p">(</span>cce12norm<span class="p">,</span> vare.mds<span class="p">,</span> type <span class="o">=</span> <span class="s">&quot;species&quot;</span><span class="p">,</span> justDF <span class="o">=</span> <span class="bp">T</span><span class="p">)</span> <span class="c1"># To get species coordinates</span>
</code></pre></div>
<p>Now we use the envfit function to fit vectors to our continuous environmental data</p>
<div class="highlight"><pre><code class="language-R" data-lang="R"><span class="kp">set.seed</span><span class="p">(</span><span class="m">152</span><span class="p">)</span>
ef <span class="o">&lt;-</span> envfit<span class="p">(</span>vare.mds<span class="p">,</span> physeqsd<span class="p">,</span> permu <span class="o">=</span> <span class="m">999</span><span class="p">)</span>
ef
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">***VECTORS

        NMDS1    NMDS2     r2  Pr(&gt;r)  
Fe    0.44906 -0.89350 0.6213 0.19861  
L1    0.26147 -0.96521 0.7724 0.05139 .
L2    0.64802  0.76163 0.1941 0.70000  
L3    0.24203 -0.97027 0.4189 0.45417  
NO3  -0.54843  0.83620 0.7571 0.08194 .
PO4  -0.58947  0.80779 0.6390 0.19861  
SiO3 -0.19340  0.98112 0.3044 0.56667  
ChlA -0.28796 -0.95764 0.6359 0.20556  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
Permutation: free
Number of permutations: 720
</code></pre></div>
<p>Use the score function again, but this time to load direction of environmental vectors. Environmental vectors here are weighted by R<sup>2</sup> value</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">scrs.vct <span class="o">&lt;-</span> <span class="kp">as.data.frame</span><span class="p">(</span>scores<span class="p">(</span>ef<span class="p">,</span> display <span class="o">=</span> <span class="s">&quot;vectors&quot;</span><span class="p">))</span>
</code></pre></div>
<p>So this was kind of confusing&hellip; ef is a list of lists. to get pvals you need to selected the pvals list from the vectors list</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">vct.pvals <span class="o">&lt;-</span> ef<span class="p">[[</span><span class="s">&#39;vectors&#39;</span><span class="p">]][</span><span class="s">&#39;pvals&#39;</span><span class="p">]</span>
</code></pre></div>
<p>combine both vectors and pvals</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">scrs.vct <span class="o">&lt;-</span> <span class="kp">cbind</span><span class="p">(</span>scrs.vct<span class="p">,</span> vct.pvals<span class="p">)</span>
</code></pre></div>
<p>select only rows with a P value &lt; 0.1</p>
<div class="highlight"><pre><code class="language-R" data-lang="R">scrs.vct <span class="o">&lt;-</span> scrs.vct<span class="p">[</span>scrs.vct<span class="o">$</span>pvals <span class="o">&lt;</span> <span class="m">0.1</span><span class="p">,</span>  <span class="p">]</span>  
</code></pre></div>
<h3 id="and-finally-create-the-plot-using-ggplot">And finally, create the plot using ggplot</h3>
<div class="highlight"><pre><code class="language-R" data-lang="R">p <span class="o">&lt;-</span> ggplot<span class="p">(</span>scrs.site<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> NMDS1<span class="p">,</span> y <span class="o">=</span> NMDS2<span class="p">))</span>
p <span class="o">+</span> geom_point<span class="p">(</span>data <span class="o">=</span> scrs.site<span class="p">,</span> aes<span class="p">(</span>shape <span class="o">=</span> Fe_Treatment<span class="p">),</span> size <span class="o">=</span> <span class="m">4</span><span class="p">)</span> <span class="o">+</span> 
  coord_fixed<span class="p">()</span> <span class="o">+</span> <span class="c1">## need aspect ratio of 1!</span>
  geom_segment<span class="p">(</span>data <span class="o">=</span> scrs.vct<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> <span class="m">0</span><span class="p">,</span> xend <span class="o">=</span> NMDS1<span class="o">/</span><span class="m">5</span><span class="p">,</span> y <span class="o">=</span> <span class="m">0</span><span class="p">,</span> yend <span class="o">=</span> NMDS2<span class="o">/</span><span class="m">5</span><span class="p">),</span> arrow <span class="o">=</span> arrow<span class="p">(</span>length <span class="o">=</span> unit<span class="p">(</span><span class="m">0.25</span><span class="p">,</span> <span class="s">&quot;cm&quot;</span><span class="p">)),</span> colour <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">)</span> <span class="o">+</span>
  geom_text<span class="p">(</span>data <span class="o">=</span> scrs.vct<span class="p">,</span> aes<span class="p">(</span>x <span class="o">=</span> NMDS1<span class="o">/</span><span class="m">4.8</span><span class="p">,</span> y <span class="o">=</span> NMDS2<span class="o">/</span><span class="m">4.8</span><span class="p">,</span> label <span class="o">=</span> <span class="kp">rownames</span><span class="p">(</span>scrs.vct<span class="p">)),</span> size <span class="o">=</span> <span class="m">3</span><span class="p">)</span> <span class="o">+</span>
  geom_point<span class="p">(</span>data <span class="o">=</span> scrs.spp<span class="p">,</span> mapping <span class="o">=</span> aes<span class="p">(</span>x <span class="o">=</span> NMDS1<span class="p">,</span> y <span class="o">=</span> NMDS2<span class="p">,</span> color <span class="o">=</span> Class<span class="p">),</span> size <span class="o">=</span> <span class="m">1.5</span><span class="p">)</span>
</code></pre></div>
<p><img src="/CCE2012_MAW_microbes/data_analysis/output_123_0.svg" alt="svg"></p>

  </article>
</div>

      </div>
    </div>
  </div>

  <footer class="footer">
  <div class="p2 wrap">
    <div class="measure mt1 center">
      <small>
        ©Copyright 2014 Shane Hogle</a> (<a href="https://twitter.com/slhogle">@slhogle</a>).
      </small>
    </div>
  </div>
</footer>



</body>
</html>
