<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>Blog</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="Microbes + Biogeochemistry">
    <meta name="author" content="Shane Hogle">
    
    <link rel="canonical" href="///CCE2012_MAW_microbes/OTU_proc/workflow/">

    <!-- Custom CSS -->
    <link rel="stylesheet" href="//css/pixyll.css" type="text/css">

    <!-- Fonts -->
    <link href='//fonts.googleapis.com/css?family=Merriweather:900,900italic,300,300italic' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Lato:900,300' rel='stylesheet' type='text/css'>
    

    <!-- Open Graph -->
    <!-- From: https://github.com/mmistakes/hpstr-jekyll-theme/blob/master/_includes/head.html -->
    <meta property="og:locale" content="en_US">
    <meta property="og:type" content="article">
    <meta property="og:title" content="Blog">
    <meta property="og:description" content="Microbes + Biogeochemistry">
    <meta property="og:url" content="//CCE2012_MAW_microbes/OTU_proc/workflow/">
    <meta property="og:site_name" content="Blog">
</head>

<body class="">
  <div class="site-wrap">
    <header class="site-header px2 px-responsive">
  <div class="mt2 wrap">
    <div class="measure">
      <a href="/" class="site-title">Blog</a>
      <nav class="site-nav right">
        <a href="//about/">Shane Hogle</a>
<a href="//publications/">Publications</a>
<a href="//research/">Research</a>
<a href="//contact/">Contact</a>

      </nav>
      <div class="clearfix"></div>
      
    </div>
  </div>
</header>


    <div class="post p2 p-responsive wrap" role="main">
      <div class="measure">
        <div class="post">
  <header class="post-header">
    <h1 class="h2"></h1>
  </header>
  <article class="post-content">
  <h2 id="reference-publication">Reference publication</h2>

<p><strong>Hogle SL</strong>, Bundy RA, Barbeau KA. <em>In prep.</em> Linking phytoplankton and bacterioplankton assemblage dynamics to iron-binding ligand production in a microcosm experiment. 
___</p>

<h2 id="introduction-otu-data-analysis">Introduction: OTU data analysis</h2>
<p>The OTU processing pipeline implemented here is basically the same as the ones recommended by Robert Edgar <a href="http://drive5.com/usearch/manual/uparse_pipeline.html">here</a> and <a href="http://drive5.com/usearch/manual/upp_ill_pe.html">here.</a> To run this pipeline you will need the <a href="http://www.drive5.com/usearch/download.html">USEARCH</a> software which includes the UPARSE algorithm. This tutorial assumes a version of USEARCH of at least v8.1.1861</p>

<p>###Read quality control
Merge paired end reads, relabel each read as FILENAME.1,2,3…</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -fastq_mergepairs DNA07_S22_L001_R1_001.fastq <span class="se">\</span>
-reverse DNA7_S22_L001_R2_001.fastq <span class="se">\</span>
-relabel @ <span class="se">\</span>
-fastqout merged.fq
</code></pre>
</div>

<p>Edgar says that for paired reads <a href="http://drive5.com/usearch/manual/upp_readprep.html">length truncation is probably unnecessary.</a> Filter out reads with more than 1.0 total expected errors for all bases in the read. There is more information about the expected error prediction in USEARCH <a href="http://drive5.com/usearch/manual/exp_errs.html">here</a> and <a href="http://bioinformatics.oxfordjournals.org/content/31/21/3476">here.</a> This is more conservative threshold than in many other comparable OTU processing pipelines. Edgar recommends using the default threshold of 1.0. This command also relabels each sequence as Filt1, Filt2, Filt3…</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -fastq_filter DNA07_merged.fastq <span class="se">\</span>
-fastq_maxee 1.0 <span class="se">\</span>
-relabel Filt <span class="se">\</span>
-fastaout filtered.fa
</code></pre>
</div>

<p>Concatenate all the sample files together into one big file</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>cat <span class="k">*</span>.fa &gt; allsamples.fa
</code></pre>
</div>

<p>###Read dereplication step</p>

<p>The input sequences to ‘-cluster_otus’ must be a set of unique sequences sorted in order of decreasing abundance with size annotations in the labels. The ‘-derep_fulllength’ command finds the unique sequences and adds size annotaitions (ie roughly how many identical sequences there were for each uniq sequence).</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -derep_fulllength allsamples.fa <span class="se">\ </span>
-relabel Uniq <span class="se">\</span>
-sizeout <span class="se">\ </span>
-fastaout allsamples_derep.fa
</code></pre>
</div>

<p>###OTU clustering step
Sorting and singleton filtering are combined into the OTU clustering step. Relabel each OTU as Otu1, Otu2, Otu3…</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -cluster_otus allsamples_derep.fa <span class="se">\</span>
-minsize 2 <span class="se">\</span>
-relabel Otu <span class="se">\</span>
-otus allsamples_otus.fa
</code></pre>
</div>

<p>Additionally, in USEARCH v8.1.1861 there is an additional chimera filtering step included in the -cluster_otus command. Edgar states that it may be possible to find extra chimera sequences by running -uchime_ref to remove predicted chimeric sequences using a certified reference database. In this case we use the <a href="http://www.mothur.org/w/images/2/21/Greengenes.gold.alignment.zip">UCHIME gold standard reference.</a> It seems that the extra chimera filtering step is mostly unecessary, so we won’t be using it in this tutorial. Check the older pipeline for the commands for uchime.</p>

<p>###Taxonomy assignment</p>

<p>Interestingly, USEARCH v8.1.1861 contains a taxonomy assignment pipeline called <a href="http://drive5.com/usearch/manual/utax_algo.html">UTAX.</a>. Edgar states that “At a high level, UTAX is a k-mer based method which looks for words in common between the query sequence and reference sequences with known taxonomy. The main advantages of UTAX over previous classifiers such as the RDP Naive Bayesian Classifier (RDP) are very high speed, informative confidence values and flexible options for training on user-supplied data.” The algorithm is currently not published, but it seems pretty robust from reading about it <a href="http://drive5.com/usearch/manual/taxonomy_validation.html">here</a> and <a href="http://drive5.com/usearch/manual/tax_train.html">here.</a></p>

<p>In order to use the UTAX algorithm, you first must download the reference data for constructing the UTAX database from <a href="http://drive5.com/usearch/manual/utax_downloads.html">here.</a> The datatbase is not precompiled or trained so you must do that yourself.</p>

<p>Change directory into  ‘/rdp_16s_trainset15/fasta’</p>

<p>This command makes a UTAX formatted database from the RDP training set which is comprised of named isolate strains. For reasons not to use a full 16S database see <a href="http://drive5.com/usearch/manual/faq_utax_largedb.html">here.</a></p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -makeudb_utax 16s_ref.fa <span class="se">\</span>
-output 16s_ref.udb <span class="se">\</span>
-report 16s_report.txt
</code></pre>
</div>

<p>This next command assigns taxonomy to each of the OTUs in the ‘allsamples_otus.fa’ file generated in the OTU clustering step.</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -utax allsamples_otus.fa <span class="se">\</span>
-db 16s_ref.udb <span class="se">\</span>
-strand both <span class="se">\</span>
-fastaout allsamples_otus_tax.fa
</code></pre>
</div>

<p>###Assign reads to OTUs</p>

<p>Finally we can <a href="http://drive5.com/usearch/manual/mapreadstootus.html">map reads to OTUs</a> and make an OTU table in a variety of formats. If OTUs have been assigned a taxonomy as in the step above this information is passed along into the OTU table. Here we generate output in a generic table format and the BIOM format which is compatible with QIIME.</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>usearch -usearch_global DNAO7/DNA07_merged.fq <span class="se">\</span>
-db allsamples_otus_tax.fa <span class="se">\</span>
-strand plus <span class="se">\</span>
-id 0.97 <span class="se">\</span>
-otutabout otutab.txt <span class="se">\</span>
-biomout otutab.json
</code></pre>
</div>

<p>###Data processing steps
For the next session you’ll want the “otutab.txt” files which is what you’ll import into R (plus some metadata files). The data processing steps are <a href="/CCE2012_MAW_microbes/data_analysis/workflow">here</a></p>

  </article>
</div>

      </div>
    </div>
  </div>

  <footer class="footer">
  <div class="p2 wrap">
    <div class="measure mt1 center">
      <small>
        &copy;Copyright 
        	<script language="javascript" type="text/javascript">
			var today = new Date()
			var year = today.getFullYear()
			document.write(year)
			</script>
			 Shane L. Hogle</a> (<a href="https://twitter.com/slhogle">@slhogle</a>).
      </small>
    </div>
  </div>
</footer>



</body>
</html>
